File: src/Silero/Detector.hs
module Silero.Detector (
  VoiceDetector (..),
  SpeechSegment (..),
  detectSegments,
  createVad,
  withVad,
) where

import Control.Applicative (Applicative (liftA2))
import Control.Exception (bracket, finally)
import Control.Monad (join)
import Control.Monad.IO.Class (MonadIO (liftIO))
import Data.Int (Int32)
import Data.Vector.Storable (Storable, Vector)
import qualified Data.Vector.Storable as Vector
import Foreign (Ptr, Storable (..), free, malloc, nullPtr, peekArray, with)
import Foreign.C (CFloat (..))
import Foreign.Storable.Generic (GStorable)
import GHC.Generics (Generic)
import Silero.Model (SileroModel (..), sampleRate, withModel)
import UnliftIO (MonadUnliftIO)

data VoiceDetector = VoiceDetector
  { model :: SileroModel
  , startThreshold :: Float
  , endThreshold :: Float
  , minSpeechSamples :: Float
  , maxSpeechSamples :: Float
  , speechPadSamples :: Float
  , minSilenceSamples :: Float
  , minSilenceSamplesAtMaxSpeech :: Float
  }
  deriving (Generic, GStorable)

-- |
-- Create a **VoiceDetector**.
-- **Warning: SileroModel holds internal state and is NOT thread safe.**
createVad :: SileroModel -> VoiceDetector
createVad model =
  VoiceDetector
    { model = model
    , startThreshold = 0.5
    , endThreshold = 0.35
    , minSpeechSamples = fromIntegral sampleRate / 1000.0 * 250.0 -- 250ms.
    , maxSpeechSamples = 1.0 / 0.0
    , speechPadSamples = fromIntegral sampleRate / 1000.0 * 30.0 -- 30ms.
    , minSilenceSamples = fromIntegral sampleRate / 1000.0 * 100.0 -- 100ms.
    , minSilenceSamplesAtMaxSpeech = fromIntegral sampleRate / 1000.0 * 98.0 -- 98ms
    }

-- |
-- Create a **VoiceDetector**.
-- **Warning: SileroModel holds internal state and is NOT thread safe.**
withVad :: (MonadUnliftIO m) => (VoiceDetector -> m a) -> m a
withVad runVad = withModel (runVad . createVad)

data SpeechSegment = SpeechSegment
  { startIndex :: Int32
  , endIndex :: Int32
  , startTime :: CFloat
  , endTime :: CFloat
  }
  deriving (Show, Read, Eq, Ord, Generic, GStorable)

foreign import ccall "detector.h detect_segments"
  c_detect_segments ::
    Ptr VoiceDetector ->
    Int -> -- samplesLength
    Ptr Float -> -- samples
    Ptr Int -> -- outSegmentsLength
    Ptr (Ptr SpeechSegment) -> -- outSegments
    IO ()

-- |
-- Detect the segments where speech starts and ends.
-- This implicitly resets the model after it finishes.
detectSegments :: (MonadIO m) => VoiceDetector -> Vector Float -> m [SpeechSegment]
detectSegments vad samples =
  liftIO . with vad $ \vadPtr ->
    Vector.unsafeWith samples $ \samplesPtr ->
      withPtr @Int $ \segmentsLengthPtr ->
        withPtr @(Ptr SpeechSegment) $ \segmentsPtr -> do
          poke samplesPtr 0
          poke segmentsPtr nullPtr
          flip finally (free =<< peek segmentsPtr) $ do
            c_detect_segments
              vadPtr
              (Vector.length samples)
              samplesPtr
              segmentsLengthPtr
              segmentsPtr
            join $
              liftA2
                peekArray
                (peek segmentsLengthPtr)
                (peek segmentsPtr)
  where
    withPtr :: forall a b. (Storable a) => (Ptr a -> IO b) -> IO b
    withPtr = bracket (malloc @a) free

File: cbits/silero/detector.c
// Credits: SileroVAD's C# example:
// https://github.com/snakers4/silero-vad/blob/cb25c0c0470cf20d7f33805218831b9620dbb6a7/examples/csharp/SileroVadDetector.cs

#include "detector.h"

void reset_segment(struct SpeechSegment *segment) {
  segment->start_index = -1;
  segment->end_index = -1;
  segment->start_time = -1.0f;
  segment->end_time = -1.0f;
}

int math_min(int a, int b) { return (a < b) ? a : b; }
int math_max(int a, int b) { return (a > b) ? a : b; }

// int compare_start_index(const void *a, const void *b) {
//   return ((struct SpeechSegment *)a)->start_index -
//          ((struct SpeechSegment *)b)->start_index;
// }

void detect_segments(struct SileroModel *model, const float start_threshold,
                     const float end_threshold, const float min_speech_samples,
                     const float max_speech_samples,
                     const float speech_pad_samples,
                     const float min_silence_samples,
                     const float min_silence_samples_at_max_speech,
                     const size_t samples_length, const float *samples,
                     size_t *out_segments_length,
                     struct SpeechSegment **out_segments) {

  size_t frames_length = samples_length / WINDOW_LENGTH;
  for (int i = 0; i < frames_length; i++) {
    const float *frame = samples + i * WINDOW_LENGTH;
    float probability = detect_speech(model, frame);
    printf("probability: %f\n", probability);
    fflush(stdout);
  }

  *out_segments_length = 1;
  // TODO: don't need calloc later when actually setting the values
  *out_segments = calloc(*out_segments_length,
                         *out_segments_length * sizeof(struct SpeechSegment));
  (*out_segments[0]).start_time = 1.2345;
  (*out_segments[0]).end_index = -2345;
}
